File: public/docs/_examples/pipes/dart/lib/fetch_json_pipe.dart
// #docregion
import 'dart:convert';
import 'dart:html';

import 'package:angular2/angular2.dart';

// #docregion pipe-metadata
@Pipe(name: 'fetch', pure: false)
// #enddocregion pipe-metadata
class FetchJsonPipe extends PipeTransform {
  dynamic _cachedData;
  String _cachedUrl;

  dynamic transform(String url) {
    if (url != _cachedUrl) {
      _cachedUrl = url;
      _cachedData = null;
      HttpRequest.getString(url).then((s) {
        _cachedData = JSON.decode(s);
      });
    }
    return _cachedData;
  }
}

File: public/docs/_examples/pipes/ts/src/app/fetch-json-pipe.ts
///<reference path="./window.extension.d.ts"/>

// #docregion
import {Pipe} from 'angular2/angular2';

// #docregion pipe-metadata
@Pipe({
  name: 'fetch',
  pure: false
})
// #enddocregion pipe-metadata
export class FetchJsonPipe {
  private fetchedValue:any;
  private fetchPromise:Promise<any>;
  transform(value:string, args:string[]):any {
    if (!this.fetchPromise) {
      this.fetchPromise = window.fetch(value)
        .then(result => result.json())
        .then(json => {
          this.fetchedValue = json;
        });
    }

    return this.fetchedValue;
  }
}